cocotb/types/array.py

```diff
@@ -285,42 +336,20 @@ def __setitem__(self, item, value):
                 "indexes must be ints or slices, not {}".format(type(item).__name__)
             )
 
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, type(self)):
+            if len(self) != len(other):
+                return False
+            return all(a == b for a, b in zip(self, other))
+        return NotImplemented
+
     def __repr__(self) -> str:
         return "{}({!r}, {!r})".format(type(self).__name__, self._value, self._range)
 
-    def __concat__(self, other: "Array") -> "Array":
-        if not isinstance(other, type(self)):
-            return NotImplemented
-        return type(self)(self._value + other._value)
-
-    def __rconcat__(self, other: "Array") -> "Array":
-        if not isinstance(other, type(self)):
-            return NotImplemented
-        return type(self)(other._value + self._value)
-
-    def index(
-        self, value: Any, start: Optional[int] = None, stop: Optional[int] = None
-    ) -> int:
-        """
-        Return index of first occurrence of *value*.
-
-        Raises :exc:`ValueError` if the value is not found.
-        Search only within *start* and *stop* if given.
-        """
-        if start is not None:
-            start = self._translate_index(start)
-        else:
-            start = 0
-        if stop is not None:
-            stop = self._translate_index(stop)
-        else:
-            stop = maxsize  # same default value used by Python lists
-        idx = self._value.index(value, start, stop)
-        return self._range[idx]
-
-    def count(self, value: Any) -> int:
-        """Return number of occurrences of *value*."""
-        return self._value.count(value)
+    def __concat__(self, other: ArrayABC[S]) -> "Array[typing.Union[T, S]]":
+        if isinstance(other, ArrayABC):
+            return Array(chain(self, other))
```

Comment:
This logic probably isn't right - if the other object is `AlternativeArray`, it's not clear to me that returning `Array` vs `AlternativeArray` is the right thing to do. Take for instance how `[] + ()` is an error in python.

Reply:
I did this on purpose so the result of concatenation is always `Array`.

This a part of my trying to find a path through the mypy minefield. You might say that if `LogicArray` was a subtype of `Array` you could do an `isinstance` check and just use `Self` types here. This requires direct inheritance because mypy does not understand virtual inheritance (`ABC.register`). That's fine for `LogicArray` because the implementation will likely be inherited, but not for subclasses like `Unsigned`, `Signed`, `Sfixed`, and `Ufixed`, etc which will likely use a different implementation.

So my options have been
1. `ArrayABC` and use virtual subclasses (which is what Python does), and forsake mypy
2. Fold `ArrayABC` back into `Array` and do direct inheritance and deal with that when implementation needs to change.
3. Only support concatenation with the current type, and likely only on `Array`, `LogicArray` and `BitArray`, requiring a cast otherwise.
4. Only support concatenation with subtypes and do direct inheritance from `Array` to `LogicArray` and `BitArray`, but leave out the integer and fixed point types, so they would need to be cast to `BitArray` first (even though they are `ArrayABC[Bit]`).

*Ideally, I'd like to see
```
def concat(Array[T], Array[S]) -> Array[T | S]
def concat(Array[T], LogicArray) -> Array[T | Logic]
def concat(Array[T], BitArray) -> Array[T | Bit]
def concat(LogicArray, LogicArray) -> LogicArray
def concat(LogicArray, BitArray) -> LogicArray
def concat(BitArray, BitArray) -> BitArray
```
In all orders and with `Unsigned`, `Signed`, `Sfixed`, `Ufixed` working in place of `BitArray`. This is fairly easy to accomplish in Python, but I am struggling trying to get this working in mypy.

Reply:
What does mypy do for `List[X] + List[Y]`?

Reply:
They have tried to change it to `List[X | Y]` [in typeshed](https://github.com/python/typeshed/pull/2404) before and it caused a [bunch of errors](https://github.com/python/mypy/issues/5492) in mypy due to how mypy handles list literals. Instead of fixing the problem, they just [reverted it](https://github.com/python/typeshed/pull/2653) =/ I [opened an issue](https://github.com/python/typeshed/issues/5756) about it.

Reply:
One more thing to note is that the operations I mentioned are not technically typesafe due to the invariance of mutable collection types. But I decided it was fine, practicality over purity, because I didn't feel like introducing a mutable vs immutable version of every type (imagine `FrozenArray[T_co]`, `Array[T]`, `FrozenLogicArray[LogicT_co]`, `LogicArray[LogicT]`, and corresponding `ArrayABC[T_co]` and `MutableArrayABC[T]` ABCs) since Python does not have easily mixed-in mutability/immutability via `const` or `mut`.

Reply:
My feeling is that probably 3 is the right choice from the list above.

Reply:
Mainly because it's easiest to add implicit conversions in a future release than it is to remove them - so I'd lean towards putting them off until we have the rest of the picture in place.

Reply:
Changed it so it only works with `Array` and not any `ArrayABC`.